tsfc/kernel_interface.py

```diff
@@ -119,400 +107,66 @@ def arguments(self, arguments, indices):
-         funarg      - :class:`coffee.Decl` function argument
-         prepare     - list of COFFEE nodes to be prepended to the
-                       kernel body
-         expressions - GEM expressions referring to the argument
-                       tensor
-         finalise    - list of COFFEE nodes to be appended to the
-                       kernel body
-    """
-    from itertools import product
-    assert isinstance(interior_facet, bool)
-
-    if len(arguments) == 0:
-        # No arguments
-        funarg = coffee.Decl(SCALAR_TYPE, coffee.Symbol("A", rank=(1,)))
-        expression = gem.Indexed(gem.Variable("A", (1,)), (0,))
-
-        return funarg, [], [expression], []
-
-    elements = tuple(create_element(arg.ufl_element()) for arg in arguments)
-
-    if not interior_facet:
-        # Not an interior facet integral
-        shape = tuple(element.space_dimension() for element in elements)
-
-        funarg = coffee.Decl(SCALAR_TYPE, coffee.Symbol("A", rank=shape))
-        expression = gem.Indexed(gem.Variable("A", shape), indices)
-
-        return funarg, [], [expression], []
-
-    if not any(isinstance(element, MixedElement) for element in elements):
-        # Interior facet integral, but no vector (mixed) arguments
-        shape = tuple(2 * element.space_dimension() for element in elements)
-
-        funarg = coffee.Decl(SCALAR_TYPE, coffee.Symbol("A", rank=shape))
-        varexp = gem.Variable("A", shape)
-
-        expressions = []
-        for restrictions in product((0, 1), repeat=len(arguments)):
-            expressions.append(gem_FlexiblyIndexed(
-                varexp,
-                tuple((r * e.space_dimension(), ((i, e.space_dimension()),))
-                      for e, i, r in zip(elements, indices, restrictions))
-            ))
-
-        return funarg, [], expressions, []
-
-    # Interior facet integral + vector (mixed) argument(s)
-    shape = tuple(element.space_dimension() for element in elements)
-    funarg_shape = tuple(s * 2 for s in shape)
-    funarg = coffee.Decl(SCALAR_TYPE, coffee.Symbol("A", rank=funarg_shape))
-
-    prepare = []
-    expressions = []
-
-    references = []
-    for restrictions in product((0, 1), repeat=len(arguments)):
-        name = "A" + "".join(map(str, restrictions))
-
-        prepare.append(coffee.Decl(SCALAR_TYPE,
-                                   coffee.Symbol(name, rank=shape),
-                                   init=coffee.ArrayInit(numpy.zeros(1))))
-        expressions.append(gem.Indexed(gem.Variable(name, shape), indices))
-
-        for multiindex in numpy.ndindex(shape):
-            references.append(coffee.Symbol(name, multiindex))
-
-    restriction_shape = []
-    for e in elements:
-        if isinstance(e, MixedElement):
-            restriction_shape += [len(e.elements()),
-                                  e.elements()[0].space_dimension()]
+class KernelBuilder(KernelBuilderBase):
+    def __new__(cls, backend, *args, **kwargs):
```

Comment:
I don't like factory functions of this kind, and elaborate switch case generally smell. Fortunately, unless one has to read external representations, these are easy to avoid. In this case one could just pass in the backend module to `compute_integral` directly. I'm testing this change now.